### PR TITLE
Bug 1128595 - Sqlite db for favicons

### PR DESCRIPTION
**Pull request by wesj**
This gets favicons using a JS shim. They're then stored locally (on disk), and the url -> site mapping is stored in a local database. The disk cache is just stored as a hash of the filename.fileExtension. That means if we can lookup the url in the database, we can load the local file. I also added a join on the history table for the history API to do that lookup simultaneously.

The schema for the favicons table is
url -> String
date -> Double // date this was saved. I need to update this as you browse I guess
width -> Int
height -> Int
type -> Int // What type of tag this came from. I'm storing things like apple-touch-icon right now, and wanted to make sure we could disambiguate those

That has speed tradeoffs I'm sure. This doesn't handle loading them asynchronously into the table.

### CODE REVIEW

File: Client/Frontend/Browser/Browser.swift [hunk 1]
```diff
@@ -6,7 +6,6 @@ import Foundation
 import WebKit
 
 protocol BrowserHelper {
-    init?(browser: Browser)
```
Comment (st3fan) [inline Client/Frontend/Browser/Browser.swift]: Why did you remove this initializer? ReaderMode.swift depends on it.

File: Client/Frontend/History/HistoryViewController.swift [hunk 2]
```diff
@@ -71,6 +71,22 @@ class HistoryViewController: UITableViewController, UrlViewController {
         }
     }
 
+    func createSizedFavicon(icon: UIImage?) -> UIImage? {
```
Comment (st3fan) [inline Client/Frontend/History/HistoryViewController.swift]: Can this be `func createSizedFavicon(icon: UIImage) -> UIImage` instead? So that it is the callers responsibility to pass in a valid favicon and make decision about what to do when it does not have a favicon?

File: Client/Frontend/History/HistoryViewController.swift [hunk 3]
```diff
@@ -79,8 +95,11 @@ class HistoryViewController: UITableViewController, UrlViewController {
             if let site = hist[indexPath.row] as? Site {
                 cell.textLabel?.text = site.title
                 cell.detailTextLabel?.text = site.url
-                // cell.imageView?.image = UIImage(named: "leaf")
-
+                if let img = site.icon?.getImage(profile.files) {
+                    cell.imageView?.image = createSizedFavicon(img)
```
Comment (st3fan) [inline Client/Frontend/History/HistoryViewController.swift]: Ha you already check if `img` is `nil` or not. So there is no need for `createSizedFavicon` to take an `UIImage` optional.

File: Providers/Profile.swift [hunk 4]
```diff
@@ -24,13 +21,26 @@ class ProfileFileAccessor : FileAccessor {
...
             }
         }
-
         return path
     }
 
-    func move(src: String, dest: String) -> Bool {
-        if let f = get(src) {
-            if let f2 = get(dest) {
+    func getDir(name: String?, basePath: String?) -> String? {
```
Comment (st3fan) [inline Providers/Profile.swift]: I know this is just test cod but I think it is better style to have `func getDir(name: String?, basePath: String? = nil) -> String?` .. then you can leave out all the `basepath: nil` params where they are not used.

File: Storage/Storage/FaviconsTable.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                       "width INTEGER, " +
+                       "height INTEGER, " +
+                       "type INTEGER NOT NULL, " +
+                       "date REAL NOT NULL" }
+
+    init(files: FileAccessor) {
+        self.files = files
+    }
+
+    override func getInsertAndArgs(inout item: Favicon) -> (String, [AnyObject?])? {
```
Comment (st3fan) [inline Storage/Storage/FaviconsTable.swift]: Why does item have to be an `inout` ?

File: Storage/Storage/SQL/BrowserDB.swift [hunk 6]
```diff
@@ -14,7 +14,7 @@ public enum FilterType {
 
 public class QueryOptions {
     // A filter string to apploy to the query
-    public var filter: String? = nil
+    public var filter: AnyObject? = nil
 
     // Allows for customizing how the filter is applied (i.e. only urls or urls and titles?)
     public var filterType: FilterType = .None
```
Comment (st3fan) [inline Storage/Storage/SQL/BrowserDB.swift]: You can probably combine `filter` and `filterType` into a single `Enum` with associated values. That will make the code more strict.

File: Storage/Storage/SQL/SQLiteHistory.swift [hunk 7]
```diff
@@ -35,10 +36,32 @@ public class SQLiteHistory : History {
...
+
+        init(cursor: Cursor) {
+            self.cursor = cursor
+            super.init(status: cursor.status, msg: cursor.statusMessage)
+        }
+
+        // Collection iteration and access functions
+        public override subscript(index: Int) -> Any? {
+            get {
+                if let (site, visit) = cursor[index] as? (Site, Visit) {
```
Comment (st3fan) [inline Storage/Storage/SQL/SQLiteHistory.swift]: If you don't care about `visit` then you can do this: `if let (site, _) = cursor[index] as? (Site, Visit)`.

File: Storage/Storage/Site.swift [hunk 8]
```diff
@@ -4,12 +4,81 @@
...
+        return "\(url.hash)"
+    }
+
+    public let url: String
+    public let date: NSDate
+    public let width: Int?
+    public let height: Int?
+    public let type: IconType
+
+    public func getImage(files: FileAccessor) -> UIImage? {
```
Comment (st3fan) [inline Storage/Storage/Site.swift]: Is there a reason this is not a read-only `lazy var img: UIImage? { get { ... } }` ?

File: Client/Frontend/Browser/Browser.swift (hunk 1 above)
Comment (wesj) [inline Client/Frontend/Browser/Browser.swift]: Does it need to be defined on the protocol? I couldn't see anywhere where we had a protocol object we were initializing.

Comment (st3fan) [inline Client/Frontend/Browser/Browser.swift]: ReaderMode is an implementation that uses it. But it does not matter. I think we can leave it out.

File: Providers/Profile.swift (hunk 4 above)
Comment (wesj) [inline Providers/Profile.swift]: Sorry, its taken me a bit to come back to this. I tried to do this initially, but Swift doesn't allow default parameter values in protocols. There's no way to do this that I know of....

File: Storage/Storage/FaviconsTable.swift (hunk 5 above)
Comment (wesj) [inline Storage/Storage/FaviconsTable.swift]: This is working around Swift compiler bugs. i.e. the wrong methods get called with out. There's a Note in Generic Table about the problem.

File: Storage/Storage/Site.swift (hunk 8 above)
Comment (wesj) [inline Storage/Storage/Site.swift]: It takes an argument (the files accessor). Storing a reference to it in every favicon would break our current setup here.

File: Client/Assets/Favicons.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        var href = icons[i].href;
+        res[href] = this.selectors[selector];
+        if (!foundIcons && this.selectors[selector] == ICON) {
+          foundIcons = true
+        }
+      }
+    }
+
+    // If we didn't find anything in the page, look to see if a favicon.ico file exists for the domain
+    if (!foundIcons) {
```
Comment (wesj) [inline Client/Assets/Favicons.js]: This is probably a stylistic thing a bit, but I wanted to make sure we always returned an object (not undefined or null). It also means we don't need an if (!res) res = {} line in the loop, but its mostly the former that seems like a pain.
